Index: admin_tools.module
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/admin_tools/admin_tools.module,v
retrieving revision 1.1
diff -u -p -r1.1 admin_tools.module
--- admin_tools.module	10 Aug 2010 17:08:23 -0000	1.1
+++ admin_tools.module	7 Oct 2010 12:09:45 -0000
@@ -88,8 +88,12 @@ function admin_tools_block($op = 'list',
   else if ($op == 'view') {
     switch($delta) {
       case 'clearcache':
-        $block = array('subject' => t('Admin Tools'),
-          'content' => admin_tools_admin_content());
+        if (user_access('administer site configuration')) {
+          $block = array(
+            'subject' => t('Admin Tools'),
+            'content' => admin_tools_admin_content(),
+          );
+        }
         break;
     }
     return $block;
